### ISSUE TITLE
Possible issue in hexbin using weights

### ISSUE THREAD

**Issue by Enzoupi** [matplotlib/matplotlib#20877]
I am facing a  problem in matplotlib, using hexbin with weights (argument C=).

When I am filling the C argument, not only the ouput color changes but also the shape. Here is a very simple code snippet reproducing the behavior:

```
  npts = int(2e5)
  
  # Create a random distribution and another one as it perturbation
  x1 = np.random.randint(0,3000,npts)
  x2 = x1 + np.random.normal(loc=0,scale=75,size=npts)
  
  # Same for the second axis variable
  y1 = np.random.randint(280,310,npts)
  y2 = y1 + np.random.normal(loc=0,scale=2,size=npts)
  
  # Plot
  fig, axes = plt.subplots(2,1)
  ax = axes[0]
  ax.hexbin(y2-y1,x1,mincnt=1,extent=[-8,8,0,3000], gridsize=150) 
  #
  ax = axes[1]
  res = ax.hexbin(y2-y1,x1,C=x1,mincnt=1,extent=[-8,8,0,3000], gridsize=150)
  
  fig.savefig(f"hexbin_weights.png")
  plt.show()
```
And the corresponding output:
[![Snippet output][2]][2]

To my understanding, as the x and y coordinates of the points are identical, only the color on the two plots should change. However, it is clearly not the case. Is it another parameter that is changing without me being aware of ? Is C not doing what I expect it to do ?

Thanks in advance for any help !

  [1]: https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.hexbin.html
  [2]: https://i.stack.imgur.com/LuqQR.png

**Comment by jklymak** [comment]
Try turning mincount to 1e-6 or something else suitably small, and I think you will figure out the issue.  

I'll close because I don't think there is an error here.  Please discuss at https://discourse.matplotlib.org if you would like more explanation.  Thanks!

**Comment by Enzoupi** [comment]
Thanks for your answer ! Although, I believe there might still be something more to say. From the hexbin documentation it is said :
```
mincnt > 0, default: None
    If not None, only display cells with more than mincnt number of points in the cell.
```
Hence the mincnt should, to my understanding of the docstring, not depend on any of the variables (not `x`,`y`, or `C`). It should be the minimum number of points landing in the hexagon no ? This might be an English related problem though...

**Comment by jklymak** [comment]
Right maybe the documentation could be clearer.  Apparently when c is provided it modifies the count (or weight) of each data point.  In this case you are making those counts be less than 1, and falling below the threshold.

**Comment by RebeccaWPerry** [comment]
I looked into this more and realized it wasn't actually a documentation-only issue.  I included more details in PR #21381, but for for the same of completeness here is an example that shows even when "scaling" by 1 and reducing with sum the bug persists.

```
import numpy as np
import matplotlib.pyplot as plt

npts = int(2e4)

# Create a random distribution and another one as it perturbation
y = np.random.randint(0, 100, npts)
# Y axis variable
x = np.random.normal(loc=0, scale=2, size=npts)

# Plot
fig, axes = plt.subplots(2,1)

#C not specified
ax = axes[0]
res1 = ax.hexbin(x, y, mincnt=1, extent=[-8,8,0,100], gridsize=150)

#adding weights of 1 and specifying reduce as sum should be identical to no weights
ax = axes[1]
res2 = ax.hexbin(x, y, C=np.ones(len(x)), mincnt=1, extent=[-8,8,0,100], gridsize=150,
                 reduce_C_function = np.sum)

c1 = res1.get_array()                
c2 = res2.get_array()

print('points binned in plot1: ', sum(c1))
print('points binned in plot2: ', sum(c2))

fig.savefig(f"hexbin_weights.png")
plt.show()
```

**Comment by rcomer** [comment]
With v3.8.0 and `main`, the code from the OP gives me

![image](https://github.com/matplotlib/matplotlib/assets/10599679/bc1f3771-1bd3-477f-93b1-2e5911566825)

So I think this is fixed, probably by #26113.
